Adding Meteor Wallet to the project

lostpebble opened the pull request:
# Description

This adds Meteor Wallet to the Wallet Selector project. I've used the other modules as a template and added Meteor configuration in all the relevant places.

# Checklist:
<!-- CHECKLIST_TYPE: ALL -->
- [x] I have performed a self-review of my own code
- [x] I have made corresponding changes to the documentation (examples)
- [x] My changes generate no new warnings
<!-- /CHECKLIST_TYPE -->

# Type of change
<!-- CHECKLIST_TYPE: ONE -->
- [x] FEATURE - a PR of this type introduces a new feature.
<!-- /CHECKLIST_TYPE -->

# Breaking changes
<!-- CHECKLIST_TYPE: ONE -->
<!-- SPECIFY BREAKING CHANGES AS A ONE-LINER -->
- [x] NO BREAKING CHANGE - this PR doesn't contain any breaking changes and it's backwards compatible
<!-- /CHECKLIST_TYPE -->

---

# Documentation

I've you'd like more documentation on the Meteor SDK, you can find that [here](https://sdk-reference.meteorwallet.app/classes/MeteorWallet). It implements an interface very similar to Near Wallet, as seen inside `near-api-js`.

## Small Issues during integration

* Running `yarn` at the root does not install `@meteorwallet/sdk` into `packages/meteor-wallet`
  * We need to run `yarn` inside the package folder
* Running examples work with Meteor Wallet, only if we delete the version of `near-api-js` inside `packages/meteor-wallet/node_modules` that was installed by `@meteorwallet/sdk` - forcing the SDK to look higher in the tree and use the same `node-api-js` used in other parts of the wallet selector.
  * This is a problem with borsh serialization and duplicate packages, where the classes have to be exactly the same classes pulled from exactly the same module during serialization and transaction creation. This should not be an issue in a real-world scenario where dependencies are de-duped at the project level correctly.

An example of what the error message looks like:

![image](https://user-images.githubusercontent.com/1508863/178274141-ddcf564b-ab77-4d4a-9260-7f1cade1e318.png)

---

The wallet is live and ready for you guys to test this integration out :)

kujtimprenkuSQA reviewed (commented):
Hi @lostpebble thanks for submitting this PR.

Can you please change the base form `main` to `dev` branch?

Can you also update the main Readme file to include `meteor-wallet` too?

For the `near-api-js` errors regarding Class Action, have you tried to match the versions?
Like using the same version of near-api-js in your project to match with the version we have in this repo?

lostpebble commented on packages/meteor-wallet/package.json:
```diff
@@ -0,0 +1,7 @@
+  "name": "@near-wallet-selector/meteor-wallet",
+  "version": "5.0.0",
+  "dependencies": {
+    "@meteorwallet/sdk": "^0.0.7"
```
ah okay- good to know. Will do.

lewis-sqa commented on packages/meteor-wallet/src/lib/meteor-wallet.ts:
```diff
@@ -0,0 +1,223 @@
+    return [{ accountId }];
+  };
+
+  const transformTransactions = async (
```
It looks like we're assuming `signerId` will be the account we're signed into. While this _should_ be case, ideally we'll want to validate this and throw to avoid potentially cryptic errors

lostpebble replied:
Yea probably for the best. In this case should I just loop through each transaction and compare the signer ID to the current account ID? What about in the case where `signerId` is undefined, is that even a possibility? (Maybe I should change the type in the case that its not)

lewis-sqa replied:
So the `signerId` should default to the active `accountId` (take a look at some of the other wallets such Ledger where we use `getActiveAccount`)

lostpebble replied:
I've added a check to the transactions and it seems to work.

lewis-sqa commented on packages/meteor-wallet/src/lib/meteor-wallet.ts:
```diff
@@ -0,0 +1,223 @@
+        const actions = transaction.actions.map((action) =>
+          createAction(action)
+        );
+        const accessKey = await account.accessKeyForTransaction(
```
What happens for transactions that require a `FullAccess` key?

lostpebble replied:
Yes for now we don't allow any `FullAccess` key transactions.

lewis-sqa commented on packages/meteor-wallet/src/lib/meteor-wallet.ts:
```diff
@@ -0,0 +1,223 @@
+  };
+
+  return {
+    async signIn({ contractId, methodNames }) {
+      logger.log("MeteorWallet:signIn", {
+        contractId,
+        methodNames,
+      });
+
+      let signInRequest: IOMeteorWalletSdk_RequestSignIn_Inputs;
+
+      if (notNullEmptyArray(methodNames)) {
+        signInRequest = {
+          methods: methodNames,
+          type: EMeteorWalletSignInType.SELECTED_METHODS,
+          contract_id: contractId,
+        } as IOMeteorWalletSdk_RequestSignIn_Inputs;
+      } else {
+        signInRequest = {
+          type: EMeteorWalletSignInType.ALL_METHODS,
+          contract_id: contractId,
+        } as IOMeteorWalletSdk_RequestSignIn_Inputs;
+      }
+
+      await _state.wallet.requestSignIn(signInRequest);
+      return getAccounts();
+    },
```
We should be able to avoid the variable and type imports by approaching this slightly differently (though I must admit, I'm not sure I understand why methods can't be optional - removing the need for a `type` property):

```suggestion
    async signIn({ contractId, methodNames = [] }) {
      logger.log("MeteorWallet:signIn", {
        contractId,
        methodNames,
      });

      if (methodNames.length) {
        await _state.wallet.requestSignIn({
          methods: methodNames,
          type: EMeteorWalletSignInType.SELECTED_METHODS,
          contract_id: contractId,
        });
      } else {
        await _state.wallet.requestSignIn({
          type: EMeteorWalletSignInType.ALL_METHODS,
          contract_id: contractId,
        });
      }
  
      return getAccounts();
    },
```

lostpebble replied:
The reason for the `type` property is just to be more deterministic instead of relying purely on type checks- feels a little strange to me that an empty array equals "all methods", where it feels like it should mean "no methods" (even though that's not an option currently). I'm not sure what cases might be added in the future as well.

lostpebble replied:
Also, in the future there might be a `MeteorWalletSignInType.FULL_ACCESS` way to sign in.

lewis-sqa commented on packages/meteor-wallet/src/lib/meteor-wallet.ts:
```diff
@@ -0,0 +1,223 @@
+}: MeteorWalletParams_Injected = {}): WalletModuleFactory<InjectedWallet> {
+  return async () => {
+    return {
+      id: "meteor-wallet-injected",
```
This should be `"meteor-wallet"`

lewis-sqa commented on packages/meteor-wallet/src/lib/meteor-wallet.ts:
```diff
@@ -0,0 +1,223 @@
+      type: "injected",
+      metadata: {
+        available: true,
+        name: "Meteor",
```
This should be `"Meteor Wallet"`

lewis-sqa commented on packages/meteor-wallet/src/lib/meteor-wallet.ts:
```diff
@@ -0,0 +1,223 @@
+      id: "meteor-wallet-injected",
+      type: "injected",
+      metadata: {
+        available: true,
```
How does the SDK work? Is it simply a wrapper around `near-api-js` or does it rely on a `window` variable?

I guess the main question is: is this wallet always available as this suggests (including mobile)?

lostpebble replied:
yep, we've tested on mobile as well- it should be always available using the method we are using (postMessage communication between windows). On mobile a new tab is opened where the user signs things, which communicates using our SDK asynchronously as well, and closes when done.

It does rely on the `window` variable for its communication, but it doesn't do full redirect / callbacks.

lewis-sqa reviewed (commented):
Few comments from me 👍

lewis-sqa commented on packages/meteor-wallet/src/lib/meteor-wallet.ts:
```diff
@@ -0,0 +1,223 @@
+  params: MeteorWalletParams_Injected,
+  network: Network
+): Promise<MeteorWalletState> => {
+  const keyStore = new keyStores.BrowserLocalStorageKeyStore();
```
I realise we probably should have done this in (My) NEAR Wallet too. I didn't initially consider these utilities to be reusable like this 😬

lewis-sqa commented:
I seem to be hitting the following error when building the repo:

```
SyntaxError: Unexpected token 'export'
```

Along with this error:

```
Uncaught at /Users/lewis/development/meteor-wallet-selector/node_modules/ (meteorwallet/sdk/dist/meteor-sdk/src/index.js:1)
    at Object.compileFunction (node:vm:352:18)
    at wrapSafe (node:internal/modules/cjs/loader:1031:15)
    at Module._compile (node:internal/modules/cjs/loader:1065:27)
    at Object.Module._extensions..js (node:internal/modules/cjs/loader:1153:10)
    at Module.load (node:internal/modules/cjs/loader:981:32)
    at Function.Module._load (node:internal/modules/cjs/loader:822:12)
    at Module.require (node:internal/modules/cjs/loader:1005:19)
    at require (node:internal/modules/cjs/helpers:102:18)
    at Object.@meteorwallet/sdk (file:///Users/lewis/development/meteor-wallet-selector/dist/examples/react/.next/server/pages/index.js:664:18)
    at __webpack_require__ (file:///Users/lewis/development/meteor-wallet-selector/dist/examples/react/.next/server/webpack-runtime.js:33:42)
```

lostpebble commented:
Hi @lewis-sqa , thanks for much for the review so far!

>I seem to be hitting the following error when building the repo:
>
>SyntaxError: Unexpected token 'export'

Yea, I noticed this error popping up now after rebasing our branch onto `dev` . It seems like the React example is built differently now, perhaps using Next?

I'm in the process of trying to solve this, nothing has changed on our side- or perhaps while I was rebasing I missed something.

lostpebble commented:
I changed our module and now I'm getting this error:

```
ReferenceError: window is not defined

This error happened while generating the page. Any console logs will be displayed in the terminal window.
```

This is because of this code in our SDK:

```
const locallySetBaseUrl = window.localStorage.getItem("DEV__METEOR_WALLET_BASE_URL");

export const envConfig: IEnvConfig = {
  wallet_base_url: locallySetBaseUrl ?? WALLET_URL_PRODUCTION_BASE,
};
```

This code is always expected to execute inside the browser... but thanks to Next.js this is now throwing an error. I've run into things like this before with them- very annoying since clearly we're building for browser but window is not allowed because it executes within a Node.js context.

lostpebble commented:
> Might be worth adding a condition around Web related APIs to make this module somewhat isomorphic. We actually moved to Next.js recently to ensure Wallet Selector handles this correctly as SSR seems to have caused problems for people in the past

That's fair, better to error early. I've wrapped it with a check now.

lostpebble commented on packages/meteor-wallet/src/lib/meteor-wallet.ts:
```diff
@@ -0,0 +1,223 @@
+
+  const wallet = new MeteorWalletSdk({ near, appKeyPrefix: "near_app" });
+
+  // Cleanup up any pending keys (cancelled logins).
+  if (!wallet.isSignedIn()) {
+    await keyStore.clear();
+  }
```
Most likely not- removing it.

lostpebble commented:
I've made a lot of edits now. It should be working, as the examples worked for me. Just need to look at some small final things recommended by @lewis-sqa about `signerId` in Transactions, and the Sign In action.

lostpebble commented:
@lewis-sqa @kujtimprenkuSQA I think I've finished adding all the requested changes and verified that its working with the example. Please check and let me know if everything is up to scratch :)

lewis-sqa commented on packages/meteor-wallet/.gitignore:
```diff
@@ -0,0 +1 @@
+/node_modules
```
I think we can remove this file now

lostpebble commented:
Shucks, always the small things that slip through! 😛 Cool, think I've changed the name in all the relevant places and removed that `.gitignore` file. Hopefully we're good to go now.

lewis-sqa reviewed (approved):
This looks great. Thanks for all your efforts @lostpebble! 🎉

lostpebble commented:
Thanks so much for the reviews and helping to get this merged! Happy to be a part of the ecosystem :)

lostpebble commented:
Is there any ETA on the next version release btw? So we can get our integrators to start testing it out

janewang commented:
@lostpebble Thanks the PR, we are planning Wallet-Selector 5.1.0 on Tuesday, August 16. We are planning to add this PR and enable Meteor Wallet. Cheers

lostpebble commented:
Hi @janewang - for a minor addition like this, this seems like a very long release schedule. I also see a minor version was released in-between without our new changes.

We are trying to iterate fast, and for this we need our integrators to start using Wallet Selector so they can test out Meteor on their dapps. This PR was merged 19 days ago now- yet there is still seemingly 15 days to go until the decided release date (more than a month total to get a version with our wallet included)... Could this also have something to do with the big official change over to MyNearWallet (I read something about a big release announcement on the 17th of August)?

This is a big blocker for us, as we've pushed for Wallet Selector to be our main integration with our users. For us to take our next beta steps we really need this functionality to be live now. Please could you consider another minor version that includes our current changes?

janewang commented:
Hey @lostpebble There is definitely significant resources devoted to various initiatives. We are looking to make sure the releases we announce can go out in time. Unfortunately the team working on wallet-selector is quite short handed esp given August vacations and natural attrition. Will discuss internally on what we can do.

lostpebble commented:
Thanks @janewang - appreciate the work of everyone involved!

janewang commented:
@lostpebble 

Please kindly confirm that the follow criteria are met or can be met within the required timeline. Thank you.

Criteria for Including New Wallets for Wallet Selector
# Product Criteria:

A wallet project must have comply the following product criteria to listed on Wallet Selector. 

- Non-custodial: The user controls their funds.
- Conformity to Wallet standards: The wallet product conforms to NEAR NEP wallet standards (Injected Wallet https://github.com/near/NEPs/pull/370 and  Bridged Wallet https://github.com/near/NEPs/pull/368)
- Ease of use: The wallet product provides a usable interface for the end user. Please provide a user guide for review.
- Ability to recover accounts: The wallet product allows users to be able to recover accounts. 
- Actively maintained: The wallet is actively maintained by a team and can answer user queries. 

# Security Criteria: 

A wallet project must have the majority of these security program features in place with self-certification that remaining items will be in place within 6 months of being listed. 

Wallets shall checkbox a statement of compliance to be maintained on the wallet’s github account.  When asserting that open items will be in place within six months of being listed, target dates shall be included. The overall statement must be dates and the date shall be commensurate with the commit date.  All wallet projects must maintain compliance. If a wallet is no longer in compliance, or no longer supported, the security statement must reflect that change. The wallet project shall make verification of the following requirements as easy as possible maximizing transparency through the use of relevant links pointing toward program descriptions, audit reports, etc. for the user/researcher.

- Has a security program in place that covers or is dedicated to the wallet
- Publishes information about its security program in an easily findable place
- Conducts regular audits of wallet code, at regular intervals of less than a year or based on meaningful code changes
- Conducts regular penetration tests, both “authenticated” and “non-authenticated” upon significant code changes
- Conducts penetration tests on related infrastructure, such as databases, virtual machines, web servers, etc. 
- Remediates any critical, high, or medium findings from audits (3, 4, an 5 above) and tests in a rapid fashion, as suggested by auditors and penetration testers–with auditors validating the remediation in their reports.
- Makes such reports (audits, penetration tests) publicly available, on at least a summary level
- When making reports (7) available, wallet projects should ensure the equivalent reports appear on the security vendor’s site or simply links to the security vendor’s report. This ensures the authenticity of the audit reports.
- Conducts operational readiness reviews and testing or an equivalent process before deployment to production to ensure that code changes have not resulted in unanticipated behavior, compatibility issues, or in
- Maintains a testnet wallet
- Maintains a bug bounty program
- Implements minimal privilege and access policies with regard to supporting infrastructure
- Implements MFA and strong passwords for access to critical related systems, such as domain registration, hosting platforms, cloud platforms, etc.
- Conducts known vulnerability and vulnerable dependency checks; and remediates critical, medium, and high findings before bringing into testnet or production.
- If the wallet is a  browser “extension,” it is listed on the official extension marketplace

lostpebble commented:
Thanks for the criteria update @janewang. I've created a public repo (https://github.com/Meteor-Wallet/meteor-public) where we will put all the required documentation and things, and let you know once we're ready. We'll also be updating the wallet according to the "Injected Wallet Standard"- although I did have a few concerns about the API there which I have mentioned on the specific PR for that document (https://github.com/near/NEPs/pull/370)- it seems that it is deviating from the API given in the current version of the Wallet Selector repo.

We really need to get merged as soon as possible because not being on Wallet Selector is putting us in a tough spot with integrators- who we've been pushing "wallet selector" on as our main integration. I hope that we will be up to standard for the next release- we're a little disappointed that this criteria only landed right now, but I'm guessing this is a new process on your side that has only just been wrapped up.

lostpebble commented:
@janewang @lewis-sqa we will be implementing our new changes in accordance with the current Wallet Selector interfaces- using our internal SDK to communicate with our extension and website (depending on context, mobile, desktop etc.). As all wallets in the wallet selector currently implement those interfaces, we hope that will be fine to allow us to be integrated in the next version.
